[MERGE] event_*, website_event_*: clean ACLs and add a new group "Registration Desk"

std-odoo opened the pull request:
[IMP] event_*: clean ACL and add a new group "Registration Desk" …

Purpose
=======

Clean the ACLs related to the Event application.

Add a new group to manage the registration in the entrance of an event. This
group should not be able to modify or remove records in Event but should be
able to create and manage the registrations.

Specifications
==============

Now, there are 3 event groups

  * ``Registration Desk User``, who can manage the registrations and
    read all event-related information;
  * ``Event User`` who can create event, sponsor, ticket... His role is
    to globally handle events on a day-to-day basis;
  * ``Event Administrator`` who can create event type, sponsor type,
    ticket type... His role is to manage the way events are managed withint
    its company:

Each group implies all previous groups.

Improve front-end so that event users can now perform some daily
operations directly in website, like publishing events or tracks.

Compared to previous event users gain a lot of rules, allowing to update
records like tickets, registrations, ... Low-end event users should now
use the registration desk group.

Links
=====

Task ID-2204364
COM #57022
ENT odoo/enterprise#13984
UPG odoo/upgrade#1897

awa-odoo commented:
Hello @std-odoo 

Some generic comments after reading this before we get into line-by-line specifics:

- This commit changes A LOT of security rules yet it introduces very little tests.
  Maybe it would be nice to test the main models (events / registrations / ...) and make sure members of the registration desk are limited as they are supposed to be while event users / managers can do more things?
- The commit message is very "light", a bit more details about the new architecture of event groups (registration desk vs user vs manager) would be nice
- When the registration desk has the same rights as "everyone", is it necessary to specify an extra line in the ACLs?
  e.g: ```access_event_track_sponsor_type_all,event.track.sponsor.type.all,model_event_sponsor_type,,1,0,0,0
access_event_track_sponsor_type_registration_desk,event.track.sponsor.type.registration,model_event_sponsor_type,event.group_event_registration_desk,1,0,0,0```
- Some choices are a bit questionable, why do they need to be able to view event.lead.rules?
  ```access_event_crm_registration,event.crm.registration,model_event_lead_rule,event.group_event_registration_desk,1,0,0,0```
- And finally, on the functional side, shouldn't event *users* only be able to modify their own events as we often do with users VS managers? (as in crm, social, ...).
  Maybe something to ask the PO

Thanks for looking into it :)

std-odoo commented:
@awa-odoo Hi :) 

I added some unit tests for the main models (+ multi-company) and I also added some explanation in the PR/commit about the 3 event grops (event user vs registration desk vs administrator).

About `event.lead.rules` yes indeed, we can drop the read access and sudo read them when needed.

> When the registration desk has the same rights as "everyone", is it necessary to specify an extra line in the ACLs?

I did that so it's easier to change the ACLs in the future or even to update them in the backend, but I can drop the unnecessary lines (not only for the registration desk but also for all users types) if you prefer.
What do you think ?

> And finally, on the functional side, shouldn't event users only be able to modify their own events as we often do with users VS managers? (as in crm, social, ...).

About your last point, LNA said that in practice many users might manage the same event. So the only solution is to add a list of "event responsible" but it's a little overkill and will add unnecessary complexity.
(in CRM only 1 vendor manager the lead)

awa-odoo commented:
@std-odoo 

> I did that so it's easier to change the ACLs in the future or even to update them in the backend, but I can drop the unnecessary lines (not only for the registration desk but also for all users types) if you prefer.
What do you think ?

I think we can get rid of unnecessary lines to make the diff a bit lighter :)

Ok for the rest, I will take a more in-depth look when you're done!

awa-odoo commented:
@std-odoo 

After re-reading, your reorganization of the csv files adds a lot of noise within the diff and it's a bit hard to track what you added/changed/removed.

Do you think you could avoid making this many changes and just add the lines where the need to be added for the new registration desk group?

If we need such a reorganization, I feel like it should be in its own commit with a REF and no other functional changes.

Thanks :)

std-odoo commented:
@awa-odoo It's not really possible because some permission for the event user/administrator also changed :/ (and also some naming has been fixed)

std-odoo commented:
And there's no functional change, python code change are there because the ACL changes broke some actions

std-odoo commented:
@awa-odoo I finished to remove the unnecessary lines :)

tde-banana-odoo commented:
Hello,

Changes in security files are impossible to review ... maybe you could have a first commit that reorganize existing rules according to their model with as less diff as possible (group all "event", then all "tags", ...). Then second commit updating them / adding some of them. Currently there are both moves and ref which make review very hard.

Don't add void lines in files, those are not necessary.

Finally adding 0, 0, 0, 0 for "all" is probably not necessary. We use a 0, 0, 0, 0 when updating existing rules and enforcing their update; if they don't exist I don't think we should add them.

Thanks,

awa-odoo reviewed (commented):
Well I was about to say pretty much the same thing as Thibault.

Changes look LGTM but the diff is especially hard to read.
I still have to find the motivation to go through all record rules one by one.

While you're reorganizing, I think the same could be done for the tests.
One commit where you MOV them to the "full" module and a second one where you add new tests.

Thanks for making the lives of your reviewers a bit easier :D

std-odoo commented:
@tde-banana-odoo 
@awa-odoo 

I splited the commit in 3 commits (clean ACLs files, move tests and change ACLs).

I can still reduce the diff on the PR by dropping the "name cleaning" (less diff in the first commit, but we will still have inconsistent names between the events modules). Do you want me to do that ?

tde-banana-odoo commented:
It's ok to improve naming. Just remember it implies migration script if you modify their xml ids :D .

std-odoo commented:
@tde-banana-odoo I think I'll juste clean the name and not the XMLID in that case :') (to avoid to have a huge migration script ^^)

std-odoo commented:
@tde-banana-odoo I tested without migration script, and after upgrading the event module from 14.0 to my branch, all the ACLs has been reset, are you sure it's needed for the ACLs ?

awa-odoo commented on addons/website_event_track_quiz/security/ir.model.access.csv:
```diff
@@ -1,7 +1,7 @@
-event_quiz_question_access_event_manager,event.quiz.question.access.event.manager,model_event_quiz_question,event.group_event_manager,1,1,1,1
-event_quiz_answer_access_event_user,event.quiz.answer.event.user,model_event_quiz_answer,event.group_event_user,1,0,0,0
-event_quiz_answer_access_event_manager,event.quiz.answer.event.manager,model_event_quiz_answer,event.group_event_manager,1,1,1,1
+event_quiz_access_event_registration,event.quiz.access.event.registration,model_event_quiz,event.group_event_registration_desk,1,0,0,0
+event_quiz_access_event_user,event.quiz.access.event.user,model_event_quiz,event.group_event_user,1,1,1,1
+event_quiz_question_access_event_registration,event.quiz.question.access.event.registration,model_event_quiz_question,event.group_event_registration_desk,1,0,0,0
+event_quiz_question_access_event_user,event.quiz.question.access.event.user,model_event_quiz_question,event.group_event_user,1,1,1,1
+event_quiz_answer_access_event_registration,event.quiz.answer.event.registration,model_event_quiz_answer,event.group_event_registration_desk,1,0,0,0
+event_quiz_answer_access_event_user,event.quiz.answer.event.user,model_event_quiz_answer,event.group_event_user,1,1,1,1
```
Well all this doesn't make too much sense.
At this point we might as well put the appropriate group on the quiz_id field on event.track so it's hidden on the form and we don't need to allow read access to all sub-models to the registration desk group.
I don't see why they would ever need it.

And arguably being able to see quiz answers is a bit of a ""private information"" even though it's not super important.

awa-odoo reviewed (commented):
Hello.
Looking a lot better since you split the commits into a MOV and a REF.
Much easier for me to see the actual changes, thanks for that!

I left some more comments about the changes, mostly on a functional point of view.
The goal is to have clean ACLs and really determine who can see what and why do they need it (well, it's kind of the whole purpose of this task I guess).

Thank you for your work :)

std-odoo commented on addons/event/security/ir.model.access.csv:
```diff
@@ -1,24 +1,26 @@
+access_event_stage_registration,event.stage.registration,model_event_stage,event.group_event_registration_desk,1,0,0,0
 access_event_stage_manager,event.stage.manager,model_event_stage,event.group_event_manager,1,1,1,1
-access_event_tag_category_user,event.tag.category.user,model_event_tag_category,event.group_event_user,1,0,0,0
+access_event_tag_category_registration,event.tag.category.registration,model_event_tag_category,event.group_event_registration_desk,1,0,0,0
```
@awa-odoo No they can read it thanks to this line (that's why I would prefer to keep necessary lines, so the ACLs are more clear)
```
access_event_tag,event.tag,model_event_tag,,1,0,0,0
```

awa-odoo replied:
My bad, I completely missed that line for some reason.

std-odoo commented on addons/website_event_track/security/ir.model.access.csv:
```diff
@@ -1,16 +1,20 @@
 id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
 access_event_track_manager,event.track.manager,model_event_track,event.group_event_manager,1,1,1,1
+access_event_track_user,event.track.user,model_event_track,event.group_event_user,1,1,1,0
```
Yes, I changed that to have less diff, but I also think it's better to have "ascendant" ACLs

std-odoo commented on addons/website_event_track/security/ir.model.access.csv:
```diff
@@ -1,16 +1,20 @@
+access_event_sponsor_user,event.sponsor.user,model_event_sponsor,event.group_event_user,1,1,1,0
+access_event_sponsor_manager,event.sponsor.manager,model_event_sponsor,event.group_event_manager,1,1,1,1
 access_event_track_stage,event.track.stage,model_event_track_stage,,1,0,0,0
+access_event_track_stage_user,event.track.stage.user,model_event_track_stage,event.group_event_user,1,1,1,0
```
Indeed, it's the role of the event manager

std-odoo commented on addons/event/security/ir.model.access.csv:
```diff
@@ -1,24 +1,26 @@
-access_event_mail_user,event.mail.user,model_event_mail,event.group_event_user,1,0,0,0
-access_event_mail_manager,event.mail.manager,model_event_mail,event.group_event_manager,1,1,1,1
-access_event_mail_registration_user,event.mail.registration.user,model_event_mail_registration,event.group_event_user,1,0,0,0
+access_event_registration_registration,event.registration.registration,model_event_registration,event.group_event_registration_desk,1,1,1,0
```
I'll make it achievable :) (PO approved)

std-odoo commented on addons/website_event_track/security/ir.model.access.csv:
```diff
@@ -1,16 +1,20 @@
-access_event_sponsor_manager,event.sponsor.manager,model_event_sponsor,event.group_event_manager,1,1,1,1
+access_event_sponsor_type_manager,event.sponsor.type.manager,model_event_sponsor_type,event.group_event_manager,1,1,1,1
 access_event_sponsor,event.sponsor,model_event_sponsor,,1,0,0,0
+access_event_sponsor_user,event.sponsor.user,model_event_sponsor,event.group_event_user,1,1,1,0
```
Same as above :) (achievable)

std-odoo commented on addons/event/security/ir.model.access.csv:
```diff
@@ -1,24 +1,26 @@
-access_event_type_ticket_user,event.type.ticket.user,model_event_type_ticket,event.group_event_user,1,0,0,0
+access_event_type_ticket_registration,event.type.ticket.registration,model_event_type_ticket,event.group_event_registration_desk,1,0,0,0
 access_event_type_ticket_manager,event.type.ticket.manager,model_event_type_ticket,event.group_event_manager,1,1,1,1
 access_event_event,event.event,model_event_event,,1,0,0,0
```
I moved this line to website event and make event readable by the users "registration desk" in Event :)

std-odoo commented:
@awa-odoo Hi :) Thank you for validate, I updated my code :)

awa-odoo commented on addons/website_event/security/ir.model.access.csv:
```diff
@@ -1,11 +1,8 @@
-access_website_event_menu_manager,website.event.menu.manager,model_website_event_menu,event.group_event_manager,1,1,1,1
+access_website_event_menu_user,website.event.menu.user,model_website_event_menu,event.group_event_user,1,1,1,1
 access_website_visitor_user,website.visitor.user,model_website_visitor,event.group_event_user,1,1,0,0
+access_event_event,event.event,model_event_event,,1,0,0,0
```
Maybe put that line at the beginning of the file since it impacts the "base" model?

Otherwise LGTM.
We will see what the reviewer thinks about the addition of archivable models since I'm still not sold on that solution.

tde-banana-odoo commented:
@robodoo r+ rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message
